src/data/list/basic.lean

```diff
@@ -572,6 +574,41 @@ begin
   { rw reverse_cons, exact H1 _ _ ih }
 end
 
+/-- Reassociates a list, revealing its last element. -/
+private def init_last_of_cons : ∀ (a : α) (l : list α), Σ' l' a', a :: l = l' ++ [a']
+| a [] := ⟨[], a, rfl⟩
+| a (b :: l) := let ⟨l', b', h'⟩ := init_last_of_cons b l in ⟨a :: l', b', by { congr, from h' }⟩
+
+/-- Implementation of `bidirectional_rec_on`. The `l` parameter is moved to the end to make the
+recursion easier to write. -/
+private def bidirectional_rec_on_aux {C : list α → Sort*}
```

Comment:
```suggestion
def bidirectional_rec {C : list α → Sort*}
```
This ordering of arguments is usually called `rec` instead of `rec_on`, and it's perfectly fine to keep it public